File: proNav.js
var Vec2D = (function(){
  var create = function(x, y){
    var obj = Object.create(def);
    obj.setXY(x, y);

    return obj;
  };


  var def ={
    _x: 1,
    _y: 0,

    getX: function()
    {
      return this._x;
    },

    setX: function(value)
    {
      this._x = value;
    },

    getY: function()
    {
      return this._y;
    },

    setY: function(value)
    {
      this._y = value;
    },

    setXY: function(x, y)
    {
      this._x = x;
      this._y = y;
    },

    getLength: function()
    {
      return Math.sqrt(this._x * this._x + this._y * this._y);
    },

    setLength: function(length)
    {
      var angle = this.getAngle();
      this._x = Math.cos(angle) * length;
      this._y = Math.sin(angle) * length;
    },

    getAngle: function()
    {
      return Math.atan2(this._y, this._x);
    },

    setAngle: function(angle)
    {
      var length = this.getLength();
      this._x = Math.cos(angle) * length;
      this._y = Math.sin(angle) * length;
    },

    add: function(vector)
    {
      this._x += vector.getX();
      this._y += vector.getY();
    },

    sub: function(vector)
    {
      this._x -= vector.getX();
      this._y -= vector.getY();
    },

    mul: function(value)
    {
      this._x *= value;
      this._y *= value;
    },

    div: function(value)
    {
      this._x /= value;
      this._y /= value;
    }
  };

  return {create:create};
}());


function catStrat4(cx, cy, rx, ry, Lx, Ly, ratThrustX, ratThrustY, catMaxThrust, losAngOld,  losLenOld, init4, catVelX, catVelY, ratVelX, ratVelY){

      var angle = 0;
      var nsep = 0;
      const N = 1
      var losAngleOld;
      var losAngleNew;
      var losLengthOld;
      var losLengthNew;
      var ratThrust = Vec2D.create(0, 0);
      ratThrust.setXY(ratThrustX, ratThrustY);
      var ratDrag = Vec2D.create(0,0);
      ratDrag.setXY(ratVelX * 5, ratVelY * 5);
      var direction = Math.PI * 0.5
      var catThrust = Vec2D.create(0,0);

      var W = 3

      var x = -(cx-rx);                     /* cat relative to rat in x*/
      var y = -(cy-ry);                     /* cat relative to rat in y*/
      x = (x > Lx/2) ? (x-Lx) : ((x<-Lx/2) ? (x+Lx) : x);      /* find the shorest distance vector from rat to cat */
      y = (y>Ly/2) ? (y-Ly) : ((y<-Ly/2) ? (y+Ly) : y);      /* on torus. */
      nsep = Math.sqrt((Math.pow(x,2)+Math.pow(y,2)));            /* relative separation normalized to box size */
      angle = Math.atan2(y,x);

      if (init4 == 1) {
        losAngleOld = angle
        losAngleNew = losAngleOld;
        losLengthOld = nsep
        losLengthNew = losLengthOld+1;
      }else{
        losAngleOld = losAngOld;
        losAngleNew = angle
        losLengthOld = losLenOld;
        losLengthNew = nsep
      }

      var lineOfSight_rate = losAngleNew - losAngleOld;
      var closingVelocity =  losLengthOld - losLengthNew;

      var tgo = (losLengthNew/closingVelocity);
      var v_rel_x = -(catVelX - ratVelX);
      var v_rel_y = -(catVelY - ratVelY);
      var n_x = (W/(tgo^2))*(x + tgo*v_rel_x + (tgo^2/2)*(ratThrust.getX()-ratDrag.getX()));
      var n_y = (W/(tgo^2))*(y + tgo*v_rel_y + (tgo^2/2)*(ratThrust.getY()-ratDrag.getY()));

      catThrust.setXY(n_x + catVelX * 2, n_y + catVelY * 2);



      var los_thrust = Vec2D.create(0,0);
      los_thrust.setLength(catThrust.getLength())
      los_thrust.setAngle(losAngleNew);


      catThrust.add(los_thrust);

      catThrust.setLength(catMaxThrust)


      var thrustArray = [catThrust.getX(), catThrust.getY(), angle, nsep];
      return thrustArray;
};

File: proNav.py
__all__ = ['proNav']

# Don't look below, you will not understand this Python code :) I don't.

from js2py.pyjs import *
# setting scope
var = Scope( JS_BUILTINS )
set_global_object(var)

# Code follows:
var.registers(['catStrat4', 'Vec2D'])
@Js
def PyJsHoisted_catStrat4_(cx, cy, rx, ry, Lx, Ly, ratThrustX, ratThrustY, catMaxThrust, losAngOld, losLenOld, init4, catVelX, catVelY, ratVelX, ratVelY, this, arguments, var=var):
    var = Scope({'cx':cx, 'cy':cy, 'rx':rx, 'ry':ry, 'Lx':Lx, 'Ly':Ly, 'ratThrustX':ratThrustX, 'ratThrustY':ratThrustY, 'catMaxThrust':catMaxThrust, 'losAngOld':losAngOld, 'losLenOld':losLenOld, 'init4':init4, 'catVelX':catVelX, 'catVelY':catVelY, 'ratVelX':ratVelX, 'ratVelY':ratVelY, 'this':this, 'arguments':arguments}, var)
    var.registers(['rx', 'lineOfSight_rate', 'losLengthOld', 'catThrust', 'thrustArray', 'Ly', 'tgo', 'N', 'losLenOld', 'catVelX', 'catMaxThrust', 'catVelY', 'nsep', 'ratThrustY', 'v_rel_x', 'W', 'losAngleNew', 'y', 'closingVelocity', 'n_y', 'cx', 'ratDrag', 'ratVelY', 'ratThrustX', 'init4', 'ratVelX', 'v_rel_y', 'Lx', 'angle', 'losAngleOld', 'ratThrust', 'n_x', 'losAngOld', 'ry', 'x', 'losLengthNew', 'direction', 'los_thrust', 'cy'])
    var.put('angle', Js(0.0))
    var.put('nsep', Js(0.0))
    var.put('N', Js(1.0))
    var.put('ratThrust', var.get('Vec2D').callprop('create', Js(0.0), Js(0.0)))
    var.get('ratThrust').callprop('setXY', var.get('ratThrustX'), var.get('ratThrustY'))
    var.put('ratDrag', var.get('Vec2D').callprop('create', Js(0.0), Js(0.0)))
    var.get('ratDrag').callprop('setXY', (var.get('ratVelX')*Js(5.0)), (var.get('ratVelY')*Js(5.0)))
    var.put('direction', (var.get('Math').get('PI')*Js(0.5)))
    var.put('catThrust', var.get('Vec2D').callprop('create', Js(0.0), Js(0.0)))
    var.put('W', Js(3.0))
    var.put('x', (-(var.get('cx')-var.get('rx'))))
    var.put('y', (-(var.get('cy')-var.get('ry'))))
    var.put('x', ((var.get('x')-var.get('Lx')) if (var.get('x')>(var.get('Lx')/Js(2.0))) else ((var.get('x')+var.get('Lx')) if (var.get('x')<((-var.get('Lx'))/Js(2.0))) else var.get('x'))))
    var.put('y', ((var.get('y')-var.get('Ly')) if (var.get('y')>(var.get('Ly')/Js(2.0))) else ((var.get('y')+var.get('Ly')) if (var.get('y')<((-var.get('Ly'))/Js(2.0))) else var.get('y'))))
    var.put('nsep', var.get('Math').callprop('sqrt', (var.get('Math').callprop('pow', var.get('x'), Js(2.0))+var.get('Math').callprop('pow', var.get('y'), Js(2.0)))))
    var.put('angle', var.get('Math').callprop('atan2', var.get('y'), var.get('x')))
    if (var.get('init4')==Js(1.0)):
        var.put('losAngleOld', var.get('angle'))
        var.put('losAngleNew', var.get('losAngleOld'))
        var.put('losLengthOld', var.get('nsep'))
        var.put('losLengthNew', (var.get('losLengthOld')+Js(1.0)))
    else:
        var.put('losAngleOld', var.get('losAngOld'))
        var.put('losAngleNew', var.get('angle'))
        var.put('losLengthOld', var.get('losLenOld'))
        var.put('losLengthNew', var.get('nsep'))
    var.put('lineOfSight_rate', (var.get('losAngleNew')-var.get('losAngleOld')))
    var.put('closingVelocity', (var.get('losLengthOld')-var.get('losLengthNew')))
    var.put('tgo', (var.get('losLengthNew')/var.get('closingVelocity')))
    var.put('v_rel_x', (-(var.get('catVelX')-var.get('ratVelX'))))
    var.put('v_rel_y', (-(var.get('catVelY')-var.get('ratVelY'))))
    var.put('n_x', ((var.get('W')/(var.get('tgo')^Js(2.0)))*((var.get('x')+(var.get('tgo')*var.get('v_rel_x')))+((var.get('tgo')^(Js(2.0)/Js(2.0)))*(var.get('ratThrust').callprop('getX')-var.get('ratDrag').callprop('getX'))))))
    var.put('n_y', ((var.get('W')/(var.get('tgo')^Js(2.0)))*((var.get('y')+(var.get('tgo')*var.get('v_rel_y')))+((var.get('tgo')^(Js(2.0)/Js(2.0)))*(var.get('ratThrust').callprop('getY')-var.get('ratDrag').callprop('getY'))))))
    var.get('catThrust').callprop('setXY', (var.get('n_x')+(var.get('catVelX')*Js(2.0))), (var.get('n_y')+(var.get('catVelY')*Js(2.0))))
    var.put('los_thrust', var.get('Vec2D').callprop('create', Js(0.0), Js(0.0)))
    var.get('los_thrust').callprop('setLength', var.get('catThrust').callprop('getLength'))
    var.get('los_thrust').callprop('setAngle', var.get('losAngleNew'))
    var.get('catThrust').callprop('add', var.get('los_thrust'))
    var.get('catThrust').callprop('setLength', var.get('catMaxThrust'))
    var.put('thrustArray', Js([var.get('catThrust').callprop('getX'), var.get('catThrust').callprop('getY'), var.get('angle'), var.get('nsep')]))
    return var.get('thrustArray')
PyJsHoisted_catStrat4_.func_name = 'catStrat4'
var.put('catStrat4', PyJsHoisted_catStrat4_)
@Js
def PyJs_anonymous_0_(this, arguments, var=var):
    var = Scope({'this':this, 'arguments':arguments}, var)
    var.registers(['def', 'create'])
    @Js
    def PyJs_anonymous_1_(x, y, this, arguments, var=var):
        var = Scope({'x':x, 'y':y, 'this':this, 'arguments':arguments}, var)
        var.registers(['y', 'obj', 'x'])
        var.put('obj', var.get('Object').callprop('create', var.get('def')))
        var.get('obj').callprop('setXY', var.get('x'), var.get('y'))
        return var.get('obj')
    PyJs_anonymous_1_._set_name('anonymous')
    var.put('create', PyJs_anonymous_1_)
    @Js
    def PyJs_anonymous_2_(this, arguments, var=var):
        var = Scope({'this':this, 'arguments':arguments}, var)
        var.registers([])
        return var.get(u"this").get('_x')
    PyJs_anonymous_2_._set_name('anonymous')
    @Js
    def PyJs_anonymous_3_(value, this, arguments, var=var):
        var = Scope({'value':value, 'this':this, 'arguments':arguments}, var)
        var.registers(['value'])
        var.get(u"this").put('_x', var.get('value'))
    PyJs_anonymous_3_._set_name('anonymous')
    @Js
    def PyJs_anonymous_4_(this, arguments, var=var):
        var = Scope({'this':this, 'arguments':arguments}, var)
        var.registers([])
        return var.get(u"this").get('_y')
    PyJs_anonymous_4_._set_name('anonymous')
    @Js
    def PyJs_anonymous_5_(value, this, arguments, var=var):
        var = Scope({'value':value, 'this':this, 'arguments':arguments}, var)
        var.registers(['value'])
        var.get(u"this").put('_y', var.get('value'))
    PyJs_anonymous_5_._set_name('anonymous')
    @Js
    def PyJs_anonymous_6_(x, y, this, arguments, var=var):
        var = Scope({'x':x, 'y':y, 'this':this, 'arguments':arguments}, var)
        var.registers(['y', 'x'])
        var.get(u"this").put('_x', var.get('x'))
        var.get(u"this").put('_y', var.get('y'))
    PyJs_anonymous_6_._set_name('anonymous')
    @Js
    def PyJs_anonymous_7_(this, arguments, var=var):
        var = Scope({'this':this, 'arguments':arguments}, var)
        var.registers([])
        return var.get('Math').callprop('sqrt', ((var.get(u"this").get('_x')*var.get(u"this").get('_x'))+(var.get(u"this").get('_y')*var.get(u"this").get('_y'))))
    PyJs_anonymous_7_._set_name('anonymous')
    @Js
    def PyJs_anonymous_8_(length, this, arguments, var=var):
        var = Scope({'length':length, 'this':this, 'arguments':arguments}, var)
        var.registers(['angle', 'length'])
        var.put('angle', var.get(u"this").callprop('getAngle'))
        var.get(u"this").put('_x', (var.get('Math').callprop('cos', var.get('angle'))*var.get('length')))
        var.get(u"this").put('_y', (var.get('Math').callprop('sin', var.get('angle'))*var.get('length')))
    PyJs_anonymous_8_._set_name('anonymous')
    @Js
    def PyJs_anonymous_9_(this, arguments, var=var):
        var = Scope({'this':this, 'arguments':arguments}, var)
        var.registers([])
        return var.get('Math').callprop('atan2', var.get(u"this").get('_y'), var.get(u"this").get('_x'))
    PyJs_anonymous_9_._set_name('anonymous')
    @Js
    def PyJs_anonymous_10_(angle, this, arguments, var=var):
        var = Scope({'angle':angle, 'this':this, 'arguments':arguments}, var)
        var.registers(['angle', 'length'])
        var.put('length', var.get(u"this").callprop('getLength'))
        var.get(u"this").put('_x', (var.get('Math').callprop('cos', var.get('angle'))*var.get('length')))
        var.get(u"this").put('_y', (var.get('Math').callprop('sin', var.get('angle'))*var.get('length')))
    PyJs_anonymous_10_._set_name('anonymous')
    @Js
    def PyJs_anonymous_11_(vector, this, arguments, var=var):
        var = Scope({'vector':vector, 'this':this, 'arguments':arguments}, var)
        var.registers(['vector'])
        var.get(u"this").put('_x', var.get('vector').callprop('getX'), '+')
        var.get(u"this").put('_y', var.get('vector').callprop('getY'), '+')
    PyJs_anonymous_11_._set_name('anonymous')
    @Js
    def PyJs_anonymous_12_(vector, this, arguments, var=var):
        var = Scope({'vector':vector, 'this':this, 'arguments':arguments}, var)
        var.registers(['vector'])
        var.get(u"this").put('_x', var.get('vector').callprop('getX'), '-')
        var.get(u"this").put('_y', var.get('vector').callprop('getY'), '-')
    PyJs_anonymous_12_._set_name('anonymous')
    @Js
    def PyJs_anonymous_13_(value, this, arguments, var=var):
        var = Scope({'value':value, 'this':this, 'arguments':arguments}, var)
        var.registers(['value'])
        var.get(u"this").put('_x', var.get('value'), '*')
        var.get(u"this").put('_y', var.get('value'), '*')
    PyJs_anonymous_13_._set_name('anonymous')
    @Js
    def PyJs_anonymous_14_(value, this, arguments, var=var):
        var = Scope({'value':value, 'this':this, 'arguments':arguments}, var)
        var.registers(['value'])
        var.get(u"this").put('_x', var.get('value'), '/')
        var.get(u"this").put('_y', var.get('value'), '/')
    PyJs_anonymous_14_._set_name('anonymous')
    var.put('def', Js({'_x':Js(1.0),'_y':Js(0.0),'getX':PyJs_anonymous_2_,'setX':PyJs_anonymous_3_,'getY':PyJs_anonymous_4_,'setY':PyJs_anonymous_5_,'setXY':PyJs_anonymous_6_,'getLength':PyJs_anonymous_7_,'setLength':PyJs_anonymous_8_,'getAngle':PyJs_anonymous_9_,'setAngle':PyJs_anonymous_10_,'add':PyJs_anonymous_11_,'sub':PyJs_anonymous_12_,'mul':PyJs_anonymous_13_,'div':PyJs_anonymous_14_}))
    return Js({'create':var.get('create')})
PyJs_anonymous_0_._set_name('anonymous')
var.put('Vec2D', PyJs_anonymous_0_())
# Add lib to the module scope
proNav = var.to_python()